### ISSUE TITLE
dart2js: Better dead code elimination - delete code that becomes unreachable due to optimizations

### ISSUE THREAD

**Issue by rakudrama** [dart-lang/sdk#43456]
dart2js avoids creating SSA nodes in the CFG builder after a _statement_ that throws.
A method call is considered to always throw is if the global analysis inferred return type is `Never`.

In the following, no call to `bigMethod()` is ever generated.
```dart
    methodThatAlwaysThrows();
    bigMethod();
```
If, however, we find that `methodThatAlwaysThrows` returns `Never` later, during SSA type propagation, we don't delete the following code.

- [ ] Methods with an inferred `Never` parameter are not called and should have an empty body.
- [ ] Code following a call with inferred return type `Never` should be removed.

**Comment by stereotype441** [comment]
If I'm understanding correctly, I believe https://dart-review.googlesource.com/c/sdk/+/162004 will fix this problem, because it inserts synthetic `throw`s after any expression having static type `Never`.

**Comment by rakudrama** [comment]
I don't think it [(162004)](https://dart-review.googlesource.com/c/sdk/+/162004) will help - it is the global type analysis that is finding _Never_, through interprocedural flow, and is nothing to do with regular Dart promotion.

**Comment by stereotype441** [comment]
> I don't think it [(162004)](https://dart-review.googlesource.com/c/sdk/+/162004) will help - it is the global type analysis that is finding _Never_, through interprocedural flow, and is nothing to do with regular Dart promotion.

Makes sense.  Thanks for clarifying.
